### PR TITLE
Introduced mentions plugin

### LINKED ISSUE #1703: Mentions plugin
## Are you reporting a feature request or a bug?

Feature request

## Provide detailed reproduction steps (if any)

It has been requested for some time to ship a Menitons/Autocomplete feature for CKEditor 4.

The idea is that as you type part of certain text, CKEditor displays a panel suggesting you some suggested values, so that you don't have to write it on your own, but use autocomplete instead.

It should support both synchronous and asynchronous entires lists (feeds).

### Multiple Feeds

The feature should also support **multiple feeds** at the same time.

For instance, you're using `@` to mention users, while `#` to mention ticket numbers.

### Differences from the [Autocomplete plugin](https://github.com/ckeditor/ckeditor-dev/issues/1751)

While Mentions feature is built on top of the Autocomplete feature, it's purpose is to expose as _stupid simple and minimal API_. It should be ready basically by providing list of users (entries) and be ready to use.

Autocomplete in contrast provides much more possibilities to control what's and how being matched, but at a cost of _more complex API_.

## Preview

Users interested in seeing how the feature looks today can preview it on [our Mentions feature branch preview](https://ckeditor4-preview.ckeditor.com/t/1703-b/plugins/mentions/samples/mentions.html).

## Browser Support

Same as for the [Autocomplete plugin](https://github.com/ckeditor/ckeditor-dev/issues/1751).

## Related Issues

* [Autocomplete plugin](https://github.com/ckeditor/ckeditor-dev/issues/1751)

### PR DESCRIPTION
**Pull request by jacekbogdanski**
## What is the purpose of this pull request?

New feature
## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

Introduced mentions plugin.

~Based on #1727~

Closes #1703

### CODE REVIEW

**Comment by mlewand** [comment]
Sorry, I've removed the `next` branch before rebasing all the PRs that were based on it. This PR is now rebased onto latest `major` (4.10.0).

**Comment by jacekbogdanski** [comment]
I rebased this branch into the latest `major` so it no longer depends on #1727.

File: tests/plugins/textwatcher/textwatcher.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+		},
+
+		'test check unmatches last match with failing match': function() {
+			var editor = this.editor, bot = this.editorBot,
+				textMatcher = attachTextWatcher( editor, function() {
+					return null;
+				} ),
+				spy = sinon.spy( textMatcher, 'unmatch' );
+
+			textMatcher.lastMatched = 'Lorem ipsum dolor sit amet, consectetur.'
```
Comment (mlewand) [inline tests/plugins/textwatcher/textwatcher.js]: Code style, missing semicolon.

File: tests/plugins/textwatcher/textwatcher.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+			textMatcher.check( {} );
+
+			assert.areEqual( 0, callbackCount );
+		},
+
+		'test check ignored with existing match': function() {
+			var editor = this.editor, bot = this.editorBot,
+				expectedMatch = 'Lorem ipsum dolor sit amet, consectetur.',
+				textMatcher = attachTextWatcher( editor, function() {
+					return { text: expectedMatch }
```
Comment (mlewand) [inline tests/plugins/textwatcher/textwatcher.js]: Code style, missing semicolon.

File: tests/plugins/textwatcher/textwatcher.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+			textMatcher.check( {} );
+
+			assert.isFalse( spy.calledWith( 'matched' ) );
+		},
+
+		'test check fired with new match': function() {
+			var editor = this.editor, bot = this.editorBot,
+				expectedMatch = 'Lorem ipsum dolor sit amet, consectetur.',
+				textMatcher = attachTextWatcher( editor, function() {
+					return { text: expectedMatch }
```
Comment (mlewand) [inline tests/plugins/textwatcher/textwatcher.js]: Code style, missing semicolon.

File: tests/plugins/textwatcher/textwatcher.js [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	function getKeyEvent( keyName, keyCode ) {
+		return {
+			name: keyName,
+			data: {
+				getKey: function() {
+					return keyCode;
+				}
+			}
```
Comment (mlewand) [inline tests/plugins/textwatcher/textwatcher.js]: Code style, missing semicolon.

File: tests/plugins/textwatcher/textwatcher.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
+/* bender-tags: editor */
+/* bender-ckeditor-plugins: textwatcher */
+
+( function() {
```
Comment (mlewand) [inline tests/plugins/textwatcher/textwatcher.js]: Please, use strict mode when possible. Apply this change to other tests too.

File: tests/plugins/textwatcher/textwatcher.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+		'test unmatch text on afterCommandExec': function() {
+			var editor = this.editor,
+				spy = sinon.spy( attachTextWatcher( editor ), 'unmatch' );
+
+			editor.fire( 'afterCommandExec' );
+
+			assert.isTrue( spy.calledOnce );
+		},
+
+		'test unmach text on blur': function() {
```
Comment (mlewand) [inline tests/plugins/textwatcher/textwatcher.js]: Typo unmach -> unmatch - correct in other places too.

File: tests/plugins/mentions/mentions.js [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+( function() {
+
+	bender.editor = true;
+
+	var feedData = [ 'Anna', 'Annabelle', 'John', 'Thomas' ],
+		expectedFeedData = [ 'Anna', 'Annabelle' ];
+
+	bender.test( {
+		'test array feed with match': function() {
+			var editor = this.editor, bot = this.editorBot,
```
[the remark this answers is not in the document]
Comment (mlewand) [reply tests/plugins/mentions/mentions.js]: While adding more cases to this suite I had to use more editors in this suite in c17b321ec however I added a property that makes the API compatible.

File: plugins/caretposition/plugin.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+	 * Returns the position of the caret.
+	 *
+	 * **Note:** In IE8 this method always returns the position of the real selection,
+	 * while in other browsers it will stick to the position of the range returned
+	 * by {@link #getRanges}, which may be one for a {@link #lock locked} or a {@link #fake fake} selection.
+	 *
+	 * **Note:** See {@link CKEDITOR.dom.range#getClientPosition} for information about possible issues
+	 * that may be caused by serious browser bugs.
+	 *
+	 * @returns {Object} The caret position
```
Comment (mlewand) [inline plugins/caretposition/plugin.js]: Again, missing `@since` tag. Please review the entire docs.

File: plugins/caretposition/plugin.js [hunk 9]
```diff
@@ -0,0 +1,168 @@
+/**
+ * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+'use strict';
+
+( function() {
+
+	CKEDITOR.plugins.add( 'caretposition', {} );
```
Comment (mlewand) [inline plugins/caretposition/plugin.js]: Actually now that we have `range.getClientRects()` do we need this plugin at all?

**Comment by mlewand** [comment]
I had another look on this PR, and noticed that URL encoding is not implemented. I have added a unit test for that.

File: plugins/caretposition/plugin.js (hunk 8 above)
Comment (jacekbogdanski) [reply plugins/caretposition/plugin.js]: `caretposition` will be probably removed so I will skip fixing it and its dependencies (e.g. tests).

**Comment by mlewand** [comment]
Now we need to reorganize this PR, since meanwhile we changed approach a bit. We want to extract it into following plugins:

* Autocomplete - will act as a base for mentions (#1751)
* Mentions - will be a super-simplified plugin (#1703)

This PR has already some work done for both of these plugins. Let's make this PR a resolution for #1751 (so there's a need to change `Closes` section in initial comment, title and description accordingly). `mentions` plugin should be extracted to a separate PR that would be based on  (and targeted at) this PR branch.

**Comment by mlewand** [comment]
Another thing that smells bad in the current code is the `isPanelActive` property.

https://github.com/ckeditor/ckeditor-dev/blob/28a57f6907a431d3f0f76582a4185f0fc92bd87b/plugins/autocomplete/plugin.js#L839

Documentation says:

> Whether the panel is active (i.e. can receive user input like click, key press). Can be modified by the setPanelActive method which fires the change-isPanelActive event.

This is incorrect. It's not the panel/view receiving the input but the whole autocomplete (as a feature, the controller). This is subtle, yet important difference. The best here is to simplify it to `isActive`. Docs should be updated accordingly.

**Comment by jacekbogdanski** [comment]
As @mlewand already mentioned https://github.com/ckeditor/ckeditor-dev/pull/1757#issuecomment-383870355 we want to split this PR into two different plugins:

* Autocomplete (#1751)
* Mentions (#1703)

To do this I already created `t/1751` target branch for this PR which will act as base branch for `autocomplete` plugin. I also removed `mentions` plugin from this PR to keep it clean. `mentions` plugin will be restored with `t/1703-b` branch based on `autocomplete` plugin (`t/1751`) by removing 2451cf6 commit.

With this changes we will separate `autocomplete` and `mentions` plugins.

**Pull request merged**